### PR TITLE
AuthN: Use single scheme as DefaultScheme

### PR DESCRIPTION
**Pull request by HaoK**
Note this is a breaking change, part of https://github.com/dotnet/aspnetcore/issues/42481

Automatically use a single authN scheme as the defaultScheme when only one is registered.

This can be disabled via: `AuthenticationOptions.DisableAutoDefaultScheme = true;`

### CODE REVIEW

**Comment by HaoK** [comment]
I started down the path of PostConfigureOptions but that didn't work at all, as DefaultScheme is a point in time thing.  Seems safest to just respect the value of this flag in the actual scheme provider implementation which is the actual source of truth for what schemes are valid.

File: src/Security/Authentication/test/FacebookTests.cs [hunk 1]
```diff
@@ -118,7 +118,7 @@ public async Task ChallengeWillIncludeScopeAsConfigured()
             app => app.UseAuthentication(),
             services =>
             {
-                services.AddAuthentication().AddFacebook(o =>
+                services.AddAuthentication(o => o.DisableAutoDefaultScheme = true).AddFacebook(o =>
```
Comment (HaoK) [inline src/Security/Authentication/test/FacebookTests.cs:121]: These tests are only testing very specific paths, setting the default scheme causes some of the other schemes to now have values which causes exceptions in these tests which are unrelated to what they are trying to test, easiest to just disable the new behavior

File: src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs [hunk 2]
```diff
@@ -52,9 +52,18 @@ protected AuthenticationSchemeProvider(IOptions<AuthenticationOptions> options,
...
+        if (_options.DefaultScheme != null)
+        {
+            return GetSchemeAsync(_options.DefaultScheme);
+        }
+        // If no default scheme is specified, and we only have a single scheme, use that
+        if (!_options.DisableAutoDefaultScheme && _schemes.Count == 1)
+        {
+            return GetSchemeAsync(_schemes.First().Key);
+        }
+        return Task.FromResult<AuthenticationScheme?>(null);
```
Comment (davidfowl) [inline src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:65]: Why do it this late?

Comment (HaoK) [reply src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:65]: manipulating the options directly has visible sideeffects to the world.  If we do it this way, we can ensure that this behavior change only affects the execution, it doesn't change the options at all

Comment (HaoK) [reply src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:65]: Basically this limits the change to the runtime behavior, as opposed to trying to pick a point in time which we set the default scheme, and then they could turn it off/null it out again after

Comment (HaoK) [reply src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:65]: Or to explain this another way, the logic is real consumer of the DefaultScheme setting, so we can guarantee we are doing the right thing in the main codepaths.  The problem with doing it on the options side, is you can add a scheme after and then we are broken.  Lets say we added a PostConfigure to apply AutoDefaultScheme, then the app adds their own `PostConfigure(o => o.AddScheme("YoloASuperLateScheme"))`, we'd have applied the default scheme when there are two schemes.  Doing it this way, we at least guarantee options are consistent, and in addition, we also correctly handle situations where apps dynamically add schemes, since authN schemes are not static for the app.

Comment (davidfowl) [reply src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:65]: Can we do it in a way where the check only happens once and not per call then?

Comment (HaoK) [reply src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:65]: See https://github.com/dotnet/aspnetcore/blob/main/src/Security/samples/DynamicSchemes/Controllers/AuthController.cs#L35 for the sample we have where you can add schemes dynamically

Comment (HaoK) [reply src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:65]: Sure, so cache it, and invalidate it if any schemes are added/removed

Comment (HaoK) [reply src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs]: Kinda nasty, but I think I did it correctly within the locks, and added a gross test that adds/removes a bunch of schemes for correctness

File: src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs [hunk 3]
```diff
@@ -47,14 +47,19 @@ protected AuthenticationSchemeProvider(IOptions<AuthenticationOptions> options,
 
     private readonly IDictionary<string, AuthenticationScheme> _schemes;
     private readonly List<AuthenticationScheme> _requestHandlers;
+    private AuthenticationScheme? _autoDefaultScheme;
```
Comment (Kahbazi) [inline src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:50]: Is it possible to store the scheme with Task to avoid creating it each time?
```suggestion
    private Task<AuthenticationScheme?> _autoDefaultScheme;
```

Comment (Kahbazi) [inline src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs:50]: Is it possible to store scheme with the Task in order to avoid creating it each time?
```suggestion
    private Task<AuthenticationScheme?> _autoDefaultScheme;
```

Comment (HaoK) [reply src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs]: Sure done

**Review by Tratcher — APPROVED** [review]
Please post a draft of the breaking change announcement here as a comment for review.

**Comment by HaoK** [comment]
Breaking change draft:

Previously, DefaultScheme needed to be explicitly set, even when there was only a single authentication scheme registered.  We've decided to change the behavior of the default IAuthenticationSchemeProvder implementation to use a single authentication scheme as the default, when no DefaultScheme is specified explicitly.  

This allows skipping specifying the DefaultScheme in your services configuration, i.e. `AddAuthentication("defaultScheme")` for apps that only have a single authentication scheme.

There is no change in behavior when a DefaultScheme is specified explicitly, or when more than one authentication scheme is registered. To opt out of this new behavior, set `AuthenticationOptions.DisableAutoDefaultScheme = true`

File: src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs [hunk 4]
```diff
@@ -113,4 +113,9 @@ public AddSchemeState([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes
     /// A user is considered authenticated if <see cref="ClaimsIdentity.IsAuthenticated"/> returns <see langword="true" /> for the <see cref="ClaimsPrincipal"/> associated with the HTTP request.
     /// </summary>
     public bool RequireAuthenticatedSignIn { get; set; } = true;
+
+    /// <summary>
+    /// If true, DefaultScheme will not automatically use a single registered scheme.
+    /// </summary>
+    public bool DisableAutoDefaultScheme { get; set; }
```
Comment (Tratcher) [inline src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: Would we consider making this an AppContext switch for one release to see if anybody ever needs it?

Comment (HaoK) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: Is that in practice better than just having a obscure public flag that no one uses?

Comment (Tratcher) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: The flag lives forever, even if nobody needs it. A switch we can remove if nobody uses it.

@DamianEdwards ?

Comment (DamianEdwards) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: It's true we can avoid the public API addition if it turns out nobody uses the `AppContext` switch for a release. I'm supportive of changing this to a switch instead and we can always add the option if we decide it's worthwhile.

Comment (HaoK) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: Sounds fine as long as we have a clear idea of how to make this determination in 8.0, or are we just going to remove it and see if anyone complains in 8?

Comment (Tratcher) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: We'll see if anyone complains in 7 and we have to tell them how to turn on the switch. I'd even leave the switch out of the announcement to encourage them to contact us.

Comment (davidfowl) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: Don’t leave it out if the announcement

Comment (Kahbazi) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: Just a thought, why not neither have the property nor the `AppContext` switch to save one if check! and whoever needed the old behavior could add a dummy authentication scheme which never authenticate. :confused:

Comment (HaoK) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: So one argument against the appcontext switch is that it really creates a headache testing, as our xunit tests run in parallel and it seems like setting this switch affects all the tests, so even if I set the switch via a try/finally, I'm seeing tests fail due to this.

They run fine if I run the tests sequentially, but when running the full set via dotnet test, I'll see failures.

How do we feel about just keeping the public option?

Comment (halter73) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: We can do both an internal option for testing and an appcontext switch for everyone else. For example:

https://github.com/dotnet/aspnetcore/blob/d96a100bddc72606f7417b665428411388b8ac54/src/Servers/Kestrel/Core/src/KestrelServerOptions.cs#L160-L176

Comment (HaoK) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: Cool that's not too bad then as the code will basically stay the same in this PR, just an additional app context check in the get for the options property

Comment (HaoK) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs:120]: Note I still think this is a net negative, as now we have internals visible to 3 more assemblies, when we could have just had this be a public bool

Comment (DamianEdwards) [reply src/Http/Authentication.Abstractions/src/AuthenticationOptions.cs]: I'll leave it you ya'll to decide, but I tend to agree with @HaoK in this case that it seems like a lot of trouble to avoid a single public `bool` property.

**Comment by adityamandaleeka** [comment]
Tentatively adding blog-candidate label to this. Probably at least worth including as part of a section on the various auth changes that went into Preview 7?
